Add spring and netty tests

joeyleeeeeee97 opened the pull request:
Add spring netty tests for the external test set.

- Using netty 4.1( the default branch) for now
- A full run of netty and spring tasks about 40 minutes on a 4C8G machine
http://ci.dragonwell-jdk.io/job/Test_openjdk8_dragonwell_sanity.external_x86-64_linux_netty/
http://ci.dragonwell-jdk.io/job/Test_openjdk8_dragonwell_sanity.external_x86-64_linux_spring/

smlambert commented:
Thanks @joeyleeeeeee97 - can I get you to take a look at [set_test_info in common_functions.sh](https://github.com/AdoptOpenJDK/openjdk-tests/blob/master/external/common_functions.sh#L175), that is currently what we use to generate Dockerfiles on the fly for different Docker OS bases (based on the value of the DOCKER_IMAGE variable).  I am hoping we can also employ the same approach for netty and spring.

In this way we have the _option_ to run external tests on (almost) all images that we produce (essentially any combination of supported_versions x supported_jvms x supported_os x supported_packages x supported_builds (as defined in [common_functions.shL16-30](https://github.com/AdoptOpenJDK/openjdk-tests/blob/master/external/common_functions.sh#L16-L30)).  If DOCKER_IMAGE is not defined, we use 'nightly' DOCKER_IMAGE tag which defaults to [ubuntu 20.04](https://github.com/AdoptOpenJDK/openjdk-docker/blob/master/11/jdk/ubuntu/Dockerfile.hotspot.nightly.full#L20) for the various versions/jvms.  And we can control what versions are actually executed through the playlist.

We currently do not use the Dockerfiles per each external sub-directory but have yet to remove them from the repo.  Sorry for the confusion.

joeyleeeeeee97 commented:
> Thanks @joeyleeeeeee97 - can I get you to take a look at [set_test_info in common_functions.sh](https://github.com/AdoptOpenJDK/openjdk-tests/blob/master/external/common_functions.sh#L175), that is currently what we use to generate Dockerfiles on the fly for different Docker OS bases (based on the value of the DOCKERIMAGE_TAG variable). I am hoping we can also employ the same approach for netty and spring.
> 
> In this way we have the _option_ to run external tests on (almost) all images that we produce (essentially any combination of supported_versions x supported_jvms x supported_os x supported_packages x supported_builds (as defined in [common_functions.shL16-30](https://github.com/AdoptOpenJDK/openjdk-tests/blob/master/external/common_functions.sh#L16-L30)). If DOCKERIMAGE_TAG is not defined, we use 'nightly' DOCKER_IMAGE tag which defaults to [ubuntu 20.04](https://github.com/AdoptOpenJDK/openjdk-docker/blob/master/11/jdk/ubuntu/Dockerfile.hotspot.nightly.full#L20) for the various versions/jvms. And we can control what versions are actually executed through the playlist.
> 
> We currently do not use the Dockerfiles per each external sub-directory but have yet to remove them from the repo. Sorry for the confusion.

Thanks for the information, I removed Dockerfile in my patch and added corresponding steps in set_test_info.

joeyleeeeeee97 commented:
I tried with specified BUILD_LIST 'external/spring' but got a no such image error :
external/spring : No such image: adoptopenjdk-spring-test
https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/5244/console

smlambert commented on external/netty/playlist.xml:
```diff
@@ -0,0 +1,34 @@
+	-->
+	<test>
+		<testCaseName>netty-test</testCaseName>
+		<command>docker run --rm $(EXTRA_DOCKER_ARGS) adoptopenjdk-netty-test:latest ; \
```
To pass in parameters from Jenkinsfilebase and pull information from common_functions.sh, we employ external.sh instead of calling docker commands directly.  
You can take the [derby command](https://github.com/AdoptOpenJDK/openjdk-tests/blob/master/external/derby/playlist.xml#L31-L34) as an example, and replace your dir name (netty) where derby is in the command, 
```
<command> $(TEST_ROOT)$(D)external$(D)external.sh --run --tag ${DOCKERIMAGE_TAG} --version ${JDK_VERSION} --impl ${JDK_IMPL} --dir netty --reportdst $(REPORTDIR) --docker_args "$(EXTRA_DOCKER_ARGS)" ; \
		$(TEST_STATUS); \
		$(TEST_ROOT)$(D)external$(D)external.sh --clean --tag ${DOCKERIMAGE_TAG} --version ${JDK_VERSION} --impl ${JDK_IMPL} --dir netty
		</command>
```

smlambert commented on external/netty/build.xml:
```diff
@@ -0,0 +1,38 @@
+<?xml version="1.0"?>
```
If we auto-generate the dockerfiles, the build.xml file can be simplified, you can use [derby build.xml](https://github.com/AdoptOpenJDK/openjdk-tests/blob/master/external/derby/build.xml) file as an example

smlambert commented on external/netty/dockerfile/netty-test.sh:
```diff
@@ -0,0 +1,48 @@
+set -e
+cd /netty/
+mvn clean package
+echo "Compile and execute example-test" && \
```
Can update to echo netty-test rather than example-test

sophia-guo replied:
Moving this comment to the line before `mvn clean package` would be clearer?

smlambert commented on external/spring/build.xml:
```diff
@@ -0,0 +1,38 @@
+<?xml version="1.0"?>
```
Same comment as for netty, just follow [derby](https://github.com/AdoptOpenJDK/openjdk-tests/blob/master/external/derby/build.xml) as an example

smlambert commented on external/spring/playlist.xml:
```diff
@@ -0,0 +1,34 @@
+	-->
+	<test>
+		<testCaseName>spring-test</testCaseName>
+		<command>docker run --rm $(EXTRA_DOCKER_ARGS) adoptopenjdk-spring-test:latest ; \
```
Same comment as for netty, this can be something like:
```
<command> $(TEST_ROOT)$(D)external$(D)external.sh --run --tag ${DOCKERIMAGE_TAG} --version ${JDK_VERSION} --impl ${JDK_IMPL} --dir spring --reportdst $(REPORTDIR) --docker_args "$(EXTRA_DOCKER_ARGS)" ; \
		$(TEST_STATUS); \
		$(TEST_ROOT)$(D)external$(D)external.sh --clean --tag ${DOCKERIMAGE_TAG} --version ${JDK_VERSION} --impl ${JDK_IMPL} --dir spring
		</command>
```

smlambert reviewed (commented):
Added some inline comments that will help get you back up and running (with auto-gen of Dockerfile). 
 
Apologies @joeyleeeeeee97 that we have not updated the example-test target, if you look at derby or any of the other subdirs you will see the new pattern to follow.  I will create an issue to update or remove example-test as it is not updated to auto-generate the Dockerfile like all the other applications.

sophia-guo commented on external/netty/dockerfile/netty-test.sh:
```diff
@@ -0,0 +1,48 @@
+java -version
+
+# Replace the following with the initial command lines that trigger execution of your test
+# For this example, we will simply compile and run the ExampleTest class
```
Those comments could be removed as it doesn't work for this case.

sophia-guo commented:
> I tried with specified BUILD_LIST 'external/spring' but got a no such image error :
external/spring : No such image: adoptopenjdk-spring-test

Image is not built. Including the updates in common_functions.sh build.xml will also need to be updated and playlist.xml as well.  Please see @smlambert 's review comment.

joeyleeeeeee97 commented:
@sophia-guo @smlambert  Thanks for your suggestions, please have another look.
I triggered a new run in 
http://ci.dragonwell-jdk.io/job/Test_openjdk8_dragonwell_sanity.external_x86-64_linux_spring/11/
http://ci.dragonwell-jdk.io/job/Test_openjdk8_dragonwell_sanity.external_x86-64_linux_netty/11/

smlambert commented:
Grinders:
spring: https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/5478
netty: https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/5477 (fails in Netty/Transport/Native/Unix/Common and skips remaining tests)

```
11:52:20  [[1;34mINFO[m] [1mReactor Summary for Netty 4.1.56.Final-SNAPSHOT:[m
11:52:20  [[1;34mINFO[m] 
11:52:20  [[1;34mINFO[m] Netty/Dev-Tools .................................... [1;32mSUCCESS[m [  9.684 s]
11:52:20  [[1;34mINFO[m] Netty .............................................. [1;32mSUCCESS[m [ 22.773 s]
11:52:20  [[1;34mINFO[m] Netty/Common ....................................... [1;32mSUCCESS[m [01:05 min]
11:52:20  [[1;34mINFO[m] Netty/Buffer ....................................... [1;32mSUCCESS[m [ 57.539 s]
11:52:20  [[1;34mINFO[m] Netty/Resolver ..................................... [1;32mSUCCESS[m [  4.899 s]
11:52:20  [[1;34mINFO[m] Netty/Transport .................................... [1;32mSUCCESS[m [01:01 min]
11:52:20  [[1;34mINFO[m] Netty/Codec ........................................ [1;32mSUCCESS[m [ 22.629 s]
11:52:20  [[1;34mINFO[m] Netty/Codec/DNS .................................... [1;32mSUCCESS[m [  7.231 s]
11:52:20  [[1;34mINFO[m] Netty/Codec/HAProxy ................................ [1;32mSUCCESS[m [  7.241 s]
11:52:20  [[1;34mINFO[m] Netty/Handler ...................................... [1;32mSUCCESS[m [09:33 min]
11:52:20  [[1;34mINFO[m] Netty/Codec/HTTP ................................... [1;32mSUCCESS[m [ 19.882 s]
11:52:20  [[1;34mINFO[m] Netty/Codec/HTTP2 .................................. [1;32mSUCCESS[m [ 35.734 s]
11:52:20  [[1;34mINFO[m] Netty/Codec/Memcache ............................... [1;32mSUCCESS[m [  5.453 s]
11:52:20  [[1;34mINFO[m] Netty/Codec/MQTT ................................... [1;32mSUCCESS[m [  5.883 s]
11:52:20  [[1;34mINFO[m] Netty/Codec/Redis .................................. [1;32mSUCCESS[m [  4.817 s]
11:52:20  [[1;34mINFO[m] Netty/Codec/SMTP ................................... [1;32mSUCCESS[m [  4.536 s]
11:52:20  [[1;34mINFO[m] Netty/Codec/Socks .................................. [1;32mSUCCESS[m [  6.367 s]
11:52:20  [[1;34mINFO[m] Netty/Codec/Stomp .................................. [1;32mSUCCESS[m [  4.496 s]
11:52:20  [[1;34mINFO[m] Netty/Codec/XML .................................... [1;32mSUCCESS[m [  4.387 s]
11:52:20  [[1;34mINFO[m] Netty/Handler/Proxy ................................ [1;32mSUCCESS[m [ 15.921 s]
11:52:20  [[1;34mINFO[m] Netty/Resolver/DNS ................................. [1;32mSUCCESS[m [ 53.432 s]
11:52:20  [[1;34mINFO[m] Netty/Transport/RXTX ............................... [1;32mSUCCESS[m [  2.324 s]
11:52:20  [[1;34mINFO[m] Netty/Transport/SCTP ............................... [1;32mSUCCESS[m [  2.569 s]
11:52:20  [[1;34mINFO[m] Netty/Transport/UDT ................................ [1;32mSUCCESS[m [ 20.947 s]
11:52:20  [[1;34mINFO[m] Netty/Example ...................................... [1;32mSUCCESS[m [  7.465 s]
11:52:20  [[1;34mINFO[m] Netty/Transport/Native/Unix/Common ................. [1;31mFAILURE[m [  3.772 s]
11:52:20  [[1;34mINFO[m] Netty/Testsuite .................................... [1;33mSKIPPED[m
11:52:20  [[1;34mINFO[m] Netty/Transport/Native/Unix/Common/Tests ........... [1;33mSKIPPED[m
11:52:20  [[1;34mINFO[m] Netty/Transport/Native/Epoll ....................... [1;33mSKIPPED[m
11:52:20  [[1;34mINFO[m] Netty/Transport/Native/KQueue ...................... [1;33mSKIPPED[m
11:52:20  [[1;34mINFO[m] Netty/Resolver/DNS/MacOS ........................... [1;33mSKIPPED[m
11:52:20  [[1;34mINFO[m] Netty/All-in-One ................................... [1;33mSKIPPED[m
11:52:20  [[1;34mINFO[m] Netty/Tarball ...................................... [1;33mSKIPPED[m
11:52:20  [[1;34mINFO[m] Netty/Testsuite/Autobahn ........................... [1;33mSKIPPED[m
11:52:20  [[1;34mINFO[m] Netty/Testsuite/Http2 .............................. [1;33mSKIPPED[m
11:52:20  [[1;34mINFO[m] Netty/Testsuite/OSGI ............................... [1;33mSKIPPED[m
11:52:20  [[1;34mINFO[m] Netty/Testsuite/Shading ............................ [1;33mSKIPPED[m
11:52:20  [[1;34mINFO[m] Netty/Testsuite/NativeImage ........................ [1;33mSKIPPED[m
11:52:20  [[1;34mINFO[m] Netty/Testsuite/NativeImage/Client ................. [1;33mSKIPPED[m
11:52:20  [[1;34mINFO[m] Netty/Testsuite/NativeImage/ClientRuntimeInit ...... [1;33mSKIPPED[m
11:52:20  [[1;34mINFO[m] Netty/Transport/BlockHound/Tests ................... [1;33mSKIPPED[m
11:52:20  [[1;34mINFO[m] Netty/Microbench ................................... [1;33mSKIPPED[m
11:52:20  [[1;34mINFO[m] Netty/BOM .......................................... [1;33mSKIPPED[m
11:52:20  [[1;34mINFO[m] [1m------------------------------------------------------------------------[m
11:52:20  [[1;34mINFO[m] [1;31mBUILD FAILURE[m
```

is the netty failure expected?

joeyleeeeeee97 commented:
> Grinders:
> spring: https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/5478
> netty: https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/5477 (fails in Netty/Transport/Native/Unix/Common and skips remaining tests)
> 
> ```
> 11:52:20  [�[1;34mINFO�[m] �[1mReactor Summary for Netty 4.1.56.Final-SNAPSHOT:�[m
> 11:52:20  [�[1;34mINFO�[m] 
> 11:52:20  [�[1;34mINFO�[m] Netty/Dev-Tools .................................... �[1;32mSUCCESS�[m [  9.684 s]
> 11:52:20  [�[1;34mINFO�[m] Netty .............................................. �[1;32mSUCCESS�[m [ 22.773 s]
> 11:52:20  [�[1;34mINFO�[m] Netty/Common ....................................... �[1;32mSUCCESS�[m [01:05 min]
> 11:52:20  [�[1;34mINFO�[m] Netty/Buffer ....................................... �[1;32mSUCCESS�[m [ 57.539 s]
> 11:52:20  [�[1;34mINFO�[m] Netty/Resolver ..................................... �[1;32mSUCCESS�[m [  4.899 s]
> 11:52:20  [�[1;34mINFO�[m] Netty/Transport .................................... �[1;32mSUCCESS�[m [01:01 min]
> 11:52:20  [�[1;34mINFO�[m] Netty/Codec ........................................ �[1;32mSUCCESS�[m [ 22.629 s]
> 11:52:20  [�[1;34mINFO�[m] Netty/Codec/DNS .................................... �[1;32mSUCCESS�[m [  7.231 s]
> 11:52:20  [�[1;34mINFO�[m] Netty/Codec/HAProxy ................................ �[1;32mSUCCESS�[m [  7.241 s]
> 11:52:20  [�[1;34mINFO�[m] Netty/Handler ...................................... �[1;32mSUCCESS�[m [09:33 min]
> 11:52:20  [�[1;34mINFO�[m] Netty/Codec/HTTP ................................... �[1;32mSUCCESS�[m [ 19.882 s]
> 11:52:20  [�[1;34mINFO�[m] Netty/Codec/HTTP2 .................................. �[1;32mSUCCESS�[m [ 35.734 s]
> 11:52:20  [�[1;34mINFO�[m] Netty/Codec/Memcache ............................... �[1;32mSUCCESS�[m [  5.453 s]
> 11:52:20  [�[1;34mINFO�[m] Netty/Codec/MQTT ................................... �[1;32mSUCCESS�[m [  5.883 s]
> 11:52:20  [�[1;34mINFO�[m] Netty/Codec/Redis .................................. �[1;32mSUCCESS�[m [  4.817 s]
> 11:52:20  [�[1;34mINFO�[m] Netty/Codec/SMTP ................................... �[1;32mSUCCESS�[m [  4.536 s]
> 11:52:20  [�[1;34mINFO�[m] Netty/Codec/Socks .................................. �[1;32mSUCCESS�[m [  6.367 s]
> 11:52:20  [�[1;34mINFO�[m] Netty/Codec/Stomp .................................. �[1;32mSUCCESS�[m [  4.496 s]
> 11:52:20  [�[1;34mINFO�[m] Netty/Codec/XML .................................... �[1;32mSUCCESS�[m [  4.387 s]
> 11:52:20  [�[1;34mINFO�[m] Netty/Handler/Proxy ................................ �[1;32mSUCCESS�[m [ 15.921 s]
> 11:52:20  [�[1;34mINFO�[m] Netty/Resolver/DNS ................................. �[1;32mSUCCESS�[m [ 53.432 s]
> 11:52:20  [�[1;34mINFO�[m] Netty/Transport/RXTX ............................... �[1;32mSUCCESS�[m [  2.324 s]
> 11:52:20  [�[1;34mINFO�[m] Netty/Transport/SCTP ............................... �[1;32mSUCCESS�[m [  2.569 s]
> 11:52:20  [�[1;34mINFO�[m] Netty/Transport/UDT ................................ �[1;32mSUCCESS�[m [ 20.947 s]
> 11:52:20  [�[1;34mINFO�[m] Netty/Example ...................................... �[1;32mSUCCESS�[m [  7.465 s]
> 11:52:20  [�[1;34mINFO�[m] Netty/Transport/Native/Unix/Common ................. �[1;31mFAILURE�[m [  3.772 s]
> 11:52:20  [�[1;34mINFO�[m] Netty/Testsuite .................................... �[1;33mSKIPPED�[m
> 11:52:20  [�[1;34mINFO�[m] Netty/Transport/Native/Unix/Common/Tests ........... �[1;33mSKIPPED�[m
> 11:52:20  [�[1;34mINFO�[m] Netty/Transport/Native/Epoll ....................... �[1;33mSKIPPED�[m
> 11:52:20  [�[1;34mINFO�[m] Netty/Transport/Native/KQueue ...................... �[1;33mSKIPPED�[m
> 11:52:20  [�[1;34mINFO�[m] Netty/Resolver/DNS/MacOS ........................... �[1;33mSKIPPED�[m
> 11:52:20  [�[1;34mINFO�[m] Netty/All-in-One ................................... �[1;33mSKIPPED�[m
> 11:52:20  [�[1;34mINFO�[m] Netty/Tarball ...................................... �[1;33mSKIPPED�[m
> 11:52:20  [�[1;34mINFO�[m] Netty/Testsuite/Autobahn ........................... �[1;33mSKIPPED�[m
> 11:52:20  [�[1;34mINFO�[m] Netty/Testsuite/Http2 .............................. �[1;33mSKIPPED�[m
> 11:52:20  [�[1;34mINFO�[m] Netty/Testsuite/OSGI ............................... �[1;33mSKIPPED�[m
> 11:52:20  [�[1;34mINFO�[m] Netty/Testsuite/Shading ............................ �[1;33mSKIPPED�[m
> 11:52:20  [�[1;34mINFO�[m] Netty/Testsuite/NativeImage ........................ �[1;33mSKIPPED�[m
> 11:52:20  [�[1;34mINFO�[m] Netty/Testsuite/NativeImage/Client ................. �[1;33mSKIPPED�[m
> 11:52:20  [�[1;34mINFO�[m] Netty/Testsuite/NativeImage/ClientRuntimeInit ...... �[1;33mSKIPPED�[m
> 11:52:20  [�[1;34mINFO�[m] Netty/Transport/BlockHound/Tests ................... �[1;33mSKIPPED�[m
> 11:52:20  [�[1;34mINFO�[m] Netty/Microbench ................................... �[1;33mSKIPPED�[m
> 11:52:20  [�[1;34mINFO�[m] Netty/BOM .......................................... �[1;33mSKIPPED�[m
> 11:52:20  [�[1;34mINFO�[m] �[1m------------------------------------------------------------------------�[m
> 11:52:20  [�[1;34mINFO�[m] �[1;31mBUILD FAILURE�[m
> ```
> 
> is the netty failure expected?
> 
> I guess we need to check if the autogenerated Dockerfile is equivalent to the static Dockerfile you had, or are there things missing...

Looks like I forgot to add "autoconf automake libtool make tar" into dependencies, now I added them and rebuild one.
https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/5487/console

smlambert commented on buildenv/jenkins/JenkinsfileBase:
```diff
@@ -115,6 +116,7 @@ def setupEnv() {
 def setupParallelEnv() {
 	stage('setupParallelEnv') {
 		timestamps{
+			def maxChildJobNum = 23
```
May as well push this to 25 (as an external_custom target will be added soon).

joeyleeeeeee97 commented:
https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/5502/
https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/5503/

@smlambert  Please have another look

smlambert reviewed (approved):
LGTM, thanks @joeyleeeeeee97 !
